### PR TITLE
Sqlite: close database on exit

### PR DESCRIPTION
**Pull request by rvl**
# Overview

The wallet server leaves behind [SQLite temporary files](https://www.sqlite.org/tempfiles.html) (`-wal` and `-shm`). The database connection should be closed cleanly before exit.

The PR changes `newDBLayer` into a bracketed function `withDBLayer`.

# Comments

This is WIP because the other tests and benchmarks still need to be updated to use `withDBLayer` rather than `newDBLayer`.

### CODE REVIEW

**Comment by KtorZ** [comment]
@rvl This error is something we ran into in the past already. As far as I understood, we can't really play around with raw sqlite methods because it breaks Persistent's abstraction boundaries. To work-around that in testing, we use the `close'` method from `Database.Persist.Sql` which makes sure to clean up all the pending prepared statement before closing the conn.

**Comment by rvl** [comment]
Ohh right, yes I thought `connClose :: Database.Persist.Sql.SqlBackend -> IO ()` was the correct function, but it's clearly not. Thanks, `close'` works.

I have fixed it up and added a test case. This PR is still WIP because not everything has been updated for the new API.

File: lib/core/src/Cardano/Wallet/DB/Sqlite.hs [hunk 1]
```diff
@@ -375,12 +405,18 @@ newDBLayer logConfig trace fp = do
...
         , withLock = \action ->
               ExceptT $ withMVar lock $ \() -> runExceptT action
-        })
+        } }
+
+-- | Finalize database statements and close the database connection.
+destroyDBLayer :: SqliteDBLayer s t -> IO ()
+destroyDBLayer (SqliteDBLayer {getSqlBackend, trace, dbFile}) = do
+    logDebug trace (MsgClosing dbFile)
+    close' getSqlBackend
```
Comment (paweljakubas) [inline lib/core/src/Cardano/Wallet/DB/Sqlite.hs]: :+1:

**Comment by KtorZ** [comment]
I've updated the remaining path of use regarding the Sqlite start and cleanup.

**Review by paweljakubas — APPROVED** [review]
now looks good! (modulo hlint)

**Review by KtorZ — APPROVED** [review]

**Pull request merged**